Allow subquery provider to have alternative constructors

Currently, a validation error is throw when a subquery provider with more than 1 constructor is encountered:
```
Caused by: java.lang.IllegalArgumentException: Invalid subquery provider with more than a single constructors:
```
Instead, the validation should only assert that there a zero arguments constructor exists.

--- a/entity-view/impl/pom.xml
+++ b/entity-view/impl/pom.xml
@@ -76,6 +76,11 @@
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
     
     <build>

--- a/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/SubqueryProviderHelper.java
+++ b/entity-view/impl/src/main/java/com/blazebit/persistence/view/impl/SubqueryProviderHelper.java
@@ -35,38 +35,54 @@ public class SubqueryProviderHelper {
     @SuppressWarnings("unchecked")
     public static SubqueryProviderFactory getFactory(Class<? extends SubqueryProvider> clazz) {
         Constructor<? extends SubqueryProvider>[] constructors = (Constructor<? extends SubqueryProvider>[]) clazz.getConstructors();
-        
-        if (constructors.length > 1) {
-            throw new IllegalArgumentException("Invalid subquery provider with more than a single constructors: " + clazz.getName());
-        }
-        
-        Constructor<? extends SubqueryProvider> constructor = constructors[0];
-        Annotation[][] parameterAnnotations = constructor.getParameterAnnotations();
-        int size = parameterAnnotations.length;
-        
-        if (size == 0) {
-            return new SimpleSubqueryProviderFactory(clazz);
-        }
-        
-        String[] parameterNames = new String[size];
-        Annotation[] annotations;
-        
-        for (int i = 0; i < size; i++) {
-            annotations = parameterAnnotations[i];
-            int annotationsSize = annotations.length;
-            
-            for (int j = 0; j < annotationsSize; j++) {
-                if (annotations[j].annotationType() == MappingParameter.class) {
-                    parameterNames[i] = ((MappingParameter) annotations[j]).value();
-                    break;
+        Constructor<? extends SubqueryProvider> noArgsConstructor = null;
+        ParameterizedSubqueryProviderFactory parameterizedSubqueryProviderFactory = null;
+        for (int constructorIdx = 0; constructorIdx < constructors.length; constructorIdx++) {
+            Constructor<? extends SubqueryProvider> constructor = constructors[constructorIdx];
+            Annotation[][] parameterAnnotations = constructor.getParameterAnnotations();
+            int size = parameterAnnotations.length;
+
+            if (size == 0) {
+                noArgsConstructor = constructor;
+            } else {
+                String[] parameterNames = new String[size];
+                Annotation[] annotations;
+
+                boolean hasMappingParameter = false;
+                int firstParameterIdxWithoutMappingParameter = -1;
+                for (int i = 0; i < size; i++) {
+                    annotations = parameterAnnotations[i];
+                    int annotationsSize = annotations.length;
+
+                    for (int j = 0; j < annotationsSize; j++) {
+                        if (annotations[j].annotationType() == MappingParameter.class) {
+                            hasMappingParameter = true;
+                            parameterNames[i] = ((MappingParameter) annotations[j]).value();
+                            break;
+                        }
+                    }
+
+                    if (parameterNames[i] == null) {
+                        firstParameterIdxWithoutMappingParameter = i;
+                    }
+                }
+                if (hasMappingParameter) {
+                    if (firstParameterIdxWithoutMappingParameter >= 0) {
+                        throw new IllegalArgumentException("Could not find any parameter mapping annotations on constructor parameter at index "
+                                + firstParameterIdxWithoutMappingParameter + " of subquery provider: " + clazz.getName());
+                    } else {
+                        parameterizedSubqueryProviderFactory = new ParameterizedSubqueryProviderFactory(constructor, parameterNames);
+                    }
                 }
             }
-            
-            if (parameterNames[i] == null) {
-                throw new IllegalArgumentException("Could not find any parameter mapping annotations on constructor parameter at index " + i + " of subquery provider: " + clazz.getName());
-            }
         }
-        
-        return new ParameterizedSubqueryProviderFactory(constructor, parameterNames);
+
+        if (parameterizedSubqueryProviderFactory != null) {
+            return parameterizedSubqueryProviderFactory;
+        } else if (noArgsConstructor != null) {
+            return new SimpleSubqueryProviderFactory(clazz);
+        } else {
+            throw new IllegalArgumentException("No eligible constructor exists for subquery provider: " + clazz.getName());
+        }
     }
 }

--- a/entity-view/impl/src/test/java/com/blazebit/persistence/view/impl/SubqueryProviderHelperTest.java
+++ b/entity-view/impl/src/test/java/com/blazebit/persistence/view/impl/SubqueryProviderHelperTest.java
@@ -1,0 +1,87 @@
+/*
+ * Copyright 2014 - 2020 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.blazebit.persistence.view.impl;
+
+import com.blazebit.persistence.SubqueryInitiator;
+import com.blazebit.persistence.view.MappingParameter;
+import com.blazebit.persistence.view.SubqueryProviderFactory;
+import org.assertj.core.api.ThrowableAssert;
+import org.junit.Test;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Moritz Becker
+ * @since 1.6.0
+ */
+public class SubqueryProviderHelperTest {
+
+    @Test
+    public void subqueryProviderValidMappingParameterConstructor() {
+        SubqueryProviderFactory subqueryProviderFactory = SubqueryProviderHelper.getFactory(SubqueryProviderValidMappingParameterConstructor.class);
+        assertTrue(subqueryProviderFactory instanceof ParameterizedSubqueryProviderFactory);
+    }
+
+    @Test
+    public void subqueryProviderInvalidMappingParameterConstructor() {
+        assertThatThrownBy(new ThrowableAssert.ThrowingCallable() {
+            @Override
+            public void call() {
+               SubqueryProviderHelper.getFactory(SubqueryProviderInvalidMappingParameterConstructor.class);
+            }
+        }).hasMessageStartingWith("Could not find any parameter mapping annotations on constructor parameter at index 0");
+    }
+
+    @Test
+    public void subqueryProviderNoArgsAndArgsConstructor() {
+        SubqueryProviderFactory subqueryProviderFactory = SubqueryProviderHelper.getFactory(SubqueryProviderNoArgsAndArgsConstructor.class);
+        assertTrue(subqueryProviderFactory instanceof SimpleSubqueryProviderFactory);
+    }
+
+    @Test
+    public void subqueryProviderNoArgsAndArgsAndMappingParameterConstructor() {
+        SubqueryProviderFactory subqueryProviderFactory = SubqueryProviderHelper.getFactory(SubqueryProviderNoArgsAndArgsAndMappingParameterConstructor.class);
+        assertTrue(subqueryProviderFactory instanceof ParameterizedSubqueryProviderFactory);
+    }
+
+    static class SubqueryProviderValidMappingParameterConstructor extends SubqueryProviderAdapter {
+        public SubqueryProviderValidMappingParameterConstructor(@MappingParameter("test") int p1) { }
+    }
+
+    static class SubqueryProviderInvalidMappingParameterConstructor extends SubqueryProviderAdapter {
+        public SubqueryProviderInvalidMappingParameterConstructor(int p1, @MappingParameter("test") int p2) { }
+    }
+
+    static class SubqueryProviderNoArgsAndArgsConstructor extends SubqueryProviderAdapter {
+        public SubqueryProviderNoArgsAndArgsConstructor() { }
+        public SubqueryProviderNoArgsAndArgsConstructor(int p1) { }
+    }
+
+    static class SubqueryProviderNoArgsAndArgsAndMappingParameterConstructor extends SubqueryProviderAdapter {
+        public SubqueryProviderNoArgsAndArgsAndMappingParameterConstructor() { }
+        public SubqueryProviderNoArgsAndArgsAndMappingParameterConstructor(int p1, int p2) { }
+        public SubqueryProviderNoArgsAndArgsAndMappingParameterConstructor(@MappingParameter("test") int p1) { }
+    }
+
+    static abstract class SubqueryProviderAdapter implements com.blazebit.persistence.view.SubqueryProvider {
+
+        @Override
+        public <T> T createSubquery(SubqueryInitiator<T> subqueryInitiator) {
+            return null;
+        }
+    }
+}

--- a/parent/pom.xml
+++ b/parent/pom.xml
@@ -526,7 +526,7 @@
             <dependency>
                 <groupId>org.assertj</groupId>
                 <artifactId>assertj-core</artifactId>
-                <version>3.18.0</version>
+                <version>3.18.1</version>
                 <scope>test</scope>
             </dependency>
 

